rfcs/20220201-quantization-scaling/README.md

```diff
@@ -0,0 +1,368 @@
+```
+
+## 3. Unifying scaling in long-term
+
+The motivation behind unifying scaling:
+- Remove duplications;
+- Provide an aligned support across all use cases.
+
+There are a few issues with the current scaling support:
+ - Output scale, scales and binary_mul overlap each other;
+ - RNN quantization is specific to RNN, but can be replaced by other
+ scaling mechanisms:
+   ```
+   # Scaling semantics of set_scales and set_zero_points
+   x_f32 = scale * (x_i8 - zero_point)
+
+   # Scaling semantics of RNN quantization
+   x_i8 = scale * x_f32 + zero_point
+   x_f32 = 1 / scale * (x_i8 - zero_point)
+   ```
+   This will affect application side code as user should pass `1 / scale` to
+   oneDNN comparing to the current `scale`;
+ - Built-in parameter provides very limited flexibility (no mask, no runtime).
+
+### The changes:
+
+#### New quantization scheme
+
+```
+(dst - zp_dst) * scale_dst = post_ops(OP(...))
+```
+Notes:
+- scales/zero_points are applied to their respective argument
+  - To get dst output of post_ops() should be divided(!) on scale_dst
+- if a users needs rescaling/reshifting in the post chain, they can rely on binary post-op
+- Scales will be supported for all data types
+
+
+#### A unified scaling API using mechanism `attr::set_scales`
+
+- The encoding of `DNNL_ARG_SCALE` is changed to `512` to avoid overlapping with other arguments.
+- Mapping of existing scaling to `attr::set_scales`:
+  - `attr::output_scales`
+    ``` cpp
+    primitive_attr attr;
+    attr.set_scales(DNNL_ARG_DST, mask, output_scales);
+    ```
+  - `attr::p_op::binary_mul`
+    ``` cpp
+     primitive_attr attr;
+     attr.set_scales(DNNL_ARG_DST, mask, output_scales);
+    ```
+  - `attr::rnn_data_qparams`
+    ``` cpp
+    primitive_attr attr;
+    attr.set_scales(DNNL_ARG_SRC, mask, data_scales);
+    attr.set_scales(DNNL_ARG_DST, mask, data_scales);
+    ```
+  - `attr::rnn_weights_qparams`
+    ``` cpp
+    primitive_attr attr;
+    attr.set_scales(DNNL_ARG_WEIGHTS, mask, weights_scales);
+    ```
+  - `attr::p_op::built-in parameter`
+    ``` cpp
+    post_ops p_ops;
+    p_ops.append_binary(algorithm::binary_add, summand_md);
+    p_ops.append_eltwise(algorithm::eltwise_relu, alpha, beta);
+
+    primitive_attr attr;
+    attr.set_post_ops(p_ops);
+    attr.set_scales(DNNL_ARG_ATTR_MULTIPLE_POST_OP(0) | DNNL_ARG_SRC_1,
+            /* mask */ 0, /* binary scale */ {DNNL_RUNTIME_F32_VAL});
+    attr.set_scales(DNNL_ARG_ATTR_MULTIPLE_POST_OP(1) | DNNL_ARG_DST,
+            /* mask */ 0, /* eltwise scale */ {DNNL_RUNTIME_F32_VAL});
+
+    ...
+    std::unordered_map<int, memory> args;
+    ...
+    args.insert({DNNL_ARG_ATTR_MULTIPLE_POST_OP(0) | DNNL_ARG_SRC_1 | DNNL_ARG_SCALE, p_op_0_src_1_scale_mem});
+    args.insert({DNNL_ARG_ATTR_MULTIPLE_POST_OP(1) | DNNL_ARG_DST | DNNL_ARG_SCALE, p_op_1_scale_mem});
+    p.execute(s, args);
+    ```
+ - `built-in parameter` in Sum
```

Comment:
We have also scale in sum post-op which seems to be missed in the document...

Reply:
I also wonder how API would look like for dw_post op. Could you describe it as well, please?